core/LSU_MMU_Shared.hpp

```diff
@@ -0,0 +1,125 @@
+#pragma once
+
+namespace olympia {
+
+
+    class MemoryAccessInfoPairDef;
+
+    class MemoryAccessInfo {
+    public:
+
+        // The modeler needs to alias a type called "SpartaPairDefinitionType" to the Pair Definition class of itself
+        using SpartaPairDefinitionType = MemoryAccessInfoPairDef;
+
+        enum class MMUState : std::uint32_t {
+            NO_ACCESS = 0,
+            __FIRST = NO_ACCESS,
+            MISS,
+            HIT,
+            NUM_STATES,
+            __LAST = NUM_STATES
+        };
+
+        enum class CacheState : std::uint64_t {
+            NO_ACCESS = 0,
+            __FIRST = NO_ACCESS,
+            MISS,
+            HIT,
+            NUM_STATES,
+            __LAST = NUM_STATES
+        };
+
+        MemoryAccessInfo() = delete;
+
+        MemoryAccessInfo(
+                const InstPtr &inst_ptr) :
+                ldst_inst_ptr_(inst_ptr),
+                phyAddrIsReady_(false),
+                mmu_access_state_(MMUState::NO_ACCESS),
+
+                // Construct the State object here
+                cache_access_state_(CacheState::NO_ACCESS) {}
+
+        virtual ~MemoryAccessInfo() {}
+
+        // This Inst pointer will act as our portal to the Inst class
+        // and we will use this pointer to query values from functions of Inst class
+        const InstPtr &getInstPtr() const { return ldst_inst_ptr_; }
+
+        // This is a function which will be added in the SPARTA_ADDPAIRs API.
+        uint64_t getInstUniqueID() const {
+            const InstPtr &inst_ptr = getInstPtr();
+
+            return inst_ptr == nullptr ? 0 : inst_ptr->getUniqueID();
+        }
+
+        void setPhyAddrStatus(bool isReady) { phyAddrIsReady_ = isReady; }
+
+        bool getPhyAddrStatus() const { return phyAddrIsReady_; }
+
+        const MMUState &getMMUState() const {
+            return mmu_access_state_;
+        }
+
+        void setMMUState(const MMUState &state) {
+            mmu_access_state_ = state;
+        }
+
+        const CacheState &getCacheState() const {
+            return cache_access_state_;
+        }
```

Comment:
MMUState and CacheState are enums (POD types).  Returning a reference converts them to a pointer and de dereferenced.  Makes sense to just return by value.   (It's a good chance the compiler will do that for you)